### PR TITLE
Add a removeListener method to DeviceEventEmitter for Framework consi…

### PR DESCRIPTION
**Pull request by rahuljiresal**
The Framework is inconsistent in how listeners are removed in certain classes. This issue has been discussed in https://github.com/facebook/react-native/issues/6493. 

For example, 

**DeviceEventEmitter**

```javascript
/* Current */
this.keyboardHideObserver = DeviceEventEmitter.addListener('keyboardWillHide', this.keyboardWillHide);
this.keyboardHideObserver.remove();

/* Expected (maybe in addition to the current API) */
DeviceEventEmitter.addListener('keyboardWillHide', this.keyboardWillHide);
DeviceEventEmitter.removeListener('keyboardWillHide', this.keyboardWillHide);
```
**AppStateIOS**

```javascript
AppStateIOS.addEventListener('change', this.handleAppStateChange);
AppStateIOS.removeEventListener('change', this.handleAppStateChange);
```

The API should be consistent, and preferably should allow both ways of removing the listeners.

Currently, developers who tried to use the second way of removing the listeners get an error for function not found. Due to the lack of documentation on `DeviceEventEmitter`, they are likely to use the next available option `DeviceEventEmitter.removeAllListeners(eventType)`, which causes bugs by removing all the listeners for that type. We need a method to remove a certain listener for a given event type.

### CODE REVIEW

**Comment by satya164** [comment]
cc @ide

**Comment by facebook-github-bot-9** [comment]
@fkgozali would you mind taking a look at this pull request? It's been a while since the last commit was reviewed.

**Comment by facebook-github-bot-6** [comment]
@rahuljiresal updated the pull request.

**Comment by facebook-github-bot-5** [comment]
@fkgozali would you mind taking a look at this pull request? It's been a while since the last commit was reviewed.

**Comment by facebook-github-bot-8** [comment]
@fkgozali would you mind taking a look at this pull request? It's been a while since the last commit was reviewed. I am pinging you based on the blame information for the files affected by this pull request.

**Comment by facebook-github-bot-6** [comment]
@fkgozali would you mind taking a look at this pull request? It's been a while since the last commit was reviewed. I am pinging you based on the blame information for the files affected by this pull request.

**Comment by facebook-github-bot-2** [comment]
@fkgozali would you mind taking a look at this pull request? It's been a while since the last commit was reviewed. I am pinging you based on the blame information for the files affected by this pull request.

**Comment by facebook-github-bot-8** [comment]
@fkgozali would you mind taking a look at this pull request? It's been a while since the last commit was reviewed. I am pinging you based on the blame information for the files affected by this pull request.

**Comment by facebook-github-bot-5** [comment]
@fkgozali would you mind taking a look at this pull request? It's been a while since the last commit was reviewed. I am pinging you based on the blame information for the files affected by this pull request.

**Comment by facebook-github-bot-2** [comment]
@fkgozali would you mind taking a look at this pull request? It's been a while since the last commit was reviewed. I am pinging you based on the blame information for the files affected by this pull request.

**Comment by facebook-github-bot-1** [comment]
It's been a while since the last commit was reviewed and the labels show this pull request needs review. Based on the blame information for the files in this pull request I identified fkgozali as a potential reviewer. Would you mind taking a look or mentioning someone with more context?

**Comment by facebook-github-bot-6** [comment]
It's been a while since the last commit was reviewed and the labels show this pull request needs review. Based on the blame information for the files in this pull request I identified @fkgozali as a potential reviewer. Would you mind taking a look or mentioning someone with more context?

**Comment by facebook-github-bot-2** [comment]
It's been a while since the last commit was reviewed and the labels show this pull request needs review. Based on the blame information for the files in this pull request I identified @fkgozali as a potential reviewer.

**Comment by fkgozali** [comment]
hey @rahuljiresal, sorry for the delayed response. I actually don't have much context on this file specifically. The code change seems fine at glance, but from #6493 - looks like we'd like to move towards `subscribe()` + `remove()` pattern eventually? So perhaps this commit is just a stop-gap?

cc @mkonicek to find reviewer who have more knowledge than me

**Comment by rahuljiresal** [comment]
@fkgozali Yes, till we have a consistent pattern, I think we should support both. Primarily so a user of the API doesn't use it incorrectly because of this inconsistency.

**Comment by facebook-github-bot-5** [comment]
@rahuljiresal updated the pull request.

**Comment by fkgozali** [comment]
@rahuljiresal yeah that makes sense.
@mkonicek I'll defer to you to merge this - seems like I don't have access to do so

**Comment by facebook-github-bot-8** [comment]
@fkgozali would you mind taking a look at this pull request? It's been a while since the last commit was reviewed.

**Comment by mkonicek** [comment]
@dmmiller What do you think about this one? Do you know a reviewer?

**Comment by dmmiller** [comment]
@nicklockwood You've been making changes here recently.  Does this make sense still considering your changes?

**Comment by nicklockwood** [comment]
@dmmiller it will need some rebasing in light of my changes.

Other than that, I think the new function itself is fine, as it calls through to subscription.remove(), which is still the canonical way to remove subscriptions.

File: Libraries/vendor/emitter/EventEmitter.js [hunk 1]
```diff
@@ -168,6 +168,36 @@ class EventEmitter {
...
+   * @param {function} listener - Function to invoke when the specified event is
+   *   emitted
+   *
+   * @example
+   *   emitter.removeListener('someEvent', function(message) {
+   *     console.log(message);
+   *   }); // removes the listener if already registered
+   *
+   */
+  removeListener(eventType: String, listener) {
```
Comment (nicklockwood) [inline Libraries/vendor/emitter/EventEmitter.js]: When you rebase, you'll notice that the implementation of the `EventEmitter.emit` method (on which this `removeListener` method seems to be based) has changed quite a lot. Please update to reflect that style, as the use of Object.keys to enumerate subscriptions (which is an array, not an Object) is a bit strange.

**Comment by facebook-github-bot-2** [comment]
@rahuljiresal updated the pull request.

**Comment by rahuljiresal** [comment]
@nicklockwood Changes are in. Please take a look. Thanks!

**Comment by facebook-github-bot-8** [comment]
@rahuljiresal updated the pull request.

**Comment by nicklockwood** [comment]
@facebook-github-bot shipit

**Pull request closed without merging**